### ISSUE TITLE
RFE: Improve quota management

### ISSUE THREAD

**Issue by dmedri** [systemd/systemd#7820]
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with

236

### Used distribution

arch

### Basic idea

Improve quota management with a new tool: quotactl.

Tree main stage:
1. info about the service. Is quota management enabled? Which partitions are already monitored or available (w/ specific options: journaled fs -> usrjquota, grpjquota) for that?
2. set limits for users or group. edquota and setquota are really far from modern standard. Inodes? A limit in k? Nowadays even fdisk can understand a "+200M". 
3. show the status. repquota is not ready for machines (see. https://github.com/systemd/systemd/issues/3552 #3552 @eliasp ).

Links:
- https://wiki.archlinux.org/index.php/disk_quota

**Comment by poettering** [comment]
Hmm, not sure I follow? systemd is not util-linux? I can see how exposing per-service quota would make sense, but per-user/per-group quota appears quite out of scope for a service manager, no?

**Comment by dmedri** [comment]
quotactl, could be the perfet companion for https://www.freedesktop.org/software/systemd/man/systemd-quotacheck.service.html.

**Comment by poettering** [comment]
uh, i am not sure we should be in that business... the quotacheck stuff in systemd is not replacing upstream quotacheck, it's just a way to integrate it in a compatible way with our boot process — we wrap it for that.

I am not convinced we should be in the business in adding new user utilities we otherwise have no dealings with.

**Comment by bluca** [comment]
We were discussing something similar at $work, and my colleague Tyler pointed out that nowadays other filesystems support XFS's "project quota" feature - ie, directory-tree-specific quotas. Ext4 supports it, for example.

Maybe we could use it together for StateDirectory/CacheDirectory? It would seem to make sense to me: those locations are strictly defined already, for example with the uid/gid being changed automatically when DynamicUser is enabled, so in a way it's "managed" storage already. Setting per service quotas should be easy enough to add as a feature, and it would take a further step in managing those locations, in case one wanted to ensure fairness or guarantee a minimal space is always available for each service and so on.

Thoughts?

**Comment by bluca** [comment]
> We were discussing something similar at $work, and my colleague Tyler pointed out that nowadays other filesystems support XFS's "project quota" feature - ie, directory-tree-specific quotas. Ext4 supports it, for example.
> 
> Maybe we could use it together for StateDirectory/CacheDirectory? It would seem to make sense to me: those locations are strictly defined already, for example with the uid/gid being changed automatically when DynamicUser is enabled, so in a way it's "managed" storage already. Setting per service quotas should be easy enough to add as a feature, and it would take a further step in managing those locations, in case one wanted to ensure fairness or guarantee a minimal space is always available for each service and so on.
> 
> Thoughts?

This was actually added to the project's TODO via https://github.com/systemd/systemd/commit/5962e9db5ebf01739208c7f1a33b2c5fb05af9cc

**Comment by behrmann** [comment]
> Maybe we could use it together for StateDirectory/CacheDirectory? It would seem to make sense to me: those locations are strictly defined already, for example with the uid/gid being changed automatically when DynamicUser is enabled, so in a way it's "managed" storage already. Setting per service quotas should be easy enough to add as a feature, and it would take a further step in managing those locations, in case one wanted to ensure fairness or guarantee a minimal space is always available for each service and so on.

As an addition: Having just picked up the pieces of some daemon implemented by people who insist on writing their own plain text logs filling up `/var/` on the slightest unforeseen situation, I was just looking for something. We'll probably never get everybody onboard with not managing their own plain text logs, so something like a `LogDirectoryQuota=` would be a great companion to `LogDirectory=`.

**Comment by bluca** [comment]
Yeah I suspect it will be similar on other FSes too, so we'll need to document the drawbacks
